### DISCUSSION TITLE
Tart VM Allocation on Studio Ultras

### DISCUSSION THREAD

**Discussion by hokadiri** [cirruslabs/tart#396] [Q&A] [answered]
An [M1 Mac Studio Ultra](https://www.apple.com/shop/buy-mac/mac-studio) has two 10-core M1 Max chips. If I deploy 2 tart VMs, what's the guarantee they'll each be deployed on each core? 
Does this answer change if I deploy say two 4-CPU tart VMs?

**Comment by fkorotkov** [comment]
Documentation around [`VZVirtualMachineConfiguration#cpuCount`](https://developer.apple.com/documentation/virtualization/vzvirtualmachineconfiguration/3656715-cpucount) doesn't provide any insight about how costs are picked and if there is any algorithm at all. [`memorySize` docs](https://developer.apple.com/documentation/virtualization/vzvirtualmachineconfiguration/3656721-memorysize) are much more insightful at the same time:

> This physical memory doesn’t necessarily map to the physical memory of the host computer. Instead, it’s a contiguous block of virtual memory that the host system reserves, but doesn’t allocate immediately, for the guest system. The guest operating system treats this memory as its physical memory and allocates pages from it as needed.

We don't have a device with Ultra processor on hand... I guess the only way is to experiment and try to find an empirical answer to the question by observing which cores are used when you run heavy workloads on a half Ultra chip.

**Comment by neoswallow** [comment, marked as answer]
What I'm seeing is that number of cores assigned equates to max percentage being used.
I have 2 VMs running on Ultra with each set to use 10 CPUs.
After looking at this I kicked off Azure Pipeline iOS build (that's what this VM is used for).
As I'm watching CPU panel from Activity Monitor, what I'm seeing is that all 20 CPUs are used on full throttle but at the max of 1,000%.
So when light weight processing is required, I see all 4 efficiency cores are being used.
With processes that require more oomph, I see powerful cores are used but mostly from 1st set of 10 cores (1 M1Max chip).
But for processes that would utilize all available processing power, I see that all 20 cores are on fire but cores that are fully working changes randomly.
VM process itself is utilizing more or less 1,000% at that point.
